skeletons/OCTET_STRING.c

```diff
@@ -1802,7 +1803,8 @@ OCTET_STRING_decode_aper(const asn_codec_ctx_t *opt_codec_ctx,
 			(long)csiz->effective_bits, (long)raw_len,
 			repeat ? "repeat" : "once", td->name);
 
-		if (raw_len > 2) { /* X.691 #16 NOTE 1 */
+		/* X.691 #16 NOTE 1 */
+		if (csiz->lower_bound != csiz->upper_bound || csiz->upper_bound > 2) {
```

Comment:
Is there a reason why we stop checking `st->size` and `raw_len`?

Reply:
We enter this branch if and only if the string has fixed length. So the actual length should match the length in constraints or encoding/decoding will fail elsewhere.

But let's hear what @velichkov thinks of course.

Reply:
Hi @OYTIS 

According to `X.680 #41` the `PrintableString` is a `RestrictedCharacterStringType` so `X.691 #30 Encoding the restricted character string types` applies to `PrintableString`. These rules are similar to `X.691 #17` but not identical.

Here are the whole `NOTE 1` from `#17` and `#30`
> 17 Encoding the octetstring type
NOTE – Octet strings of **fixed length less than or equal to two octets** are not octet-aligned. All other octet strings are octet-aligned in the ALIGNED variant. Fixed length octet strings encode with no length octets if they are shorter than 64K. For unconstrained octet strings the length is explicitly encoded (with fragmentation if necessary).

>30 Encoding the restricted character string types
NOTE 1 – (Tutorial ALIGNED variant) Character strings of **fixed length less than or equal to two octets** are not octet-aligned. Character strings of **variable length** that are constrained to have a maximum **length of less than two octets** are not octet-aligned. All other character strings are octet-aligned in the ALIGNED variant. Fixed length character strings encode with no length octets if they are shorter than 64K characters. For unconstrained character strings or constrained character strings longer than 64K–1, the length is explicitly encoded (with fragmentation if necessary). Each NumericString, PrintableString, VisibleString (ISO646String), IA5String, BMPString and UniversalString character is encoded into the number of bits that is the smallest power of two that can accommodate all characters allowed by the effective permitted-alphabet constraint.

Note that they are talking about **octets** and not **characters** and one character is not always one octet, for example UniversalString is defined to be `ISO/IEC 10646` where one character is 4 octets.

Note also that `#30 NOTE 1` contains at least one clause (the one about variable length less then two octets) that is not present in `#17 NOTE` so we have to take this into consideration because currently the code for both `OCTET STRING` and `PrintableString/UniversalString/...` is the same.

So I believe it's better to revert `csiz->upper_bound > 2` to `raw_len` and `std->size` respectively and make it conditional to the type that's being encoded/decoded. Right now I'm not sure how to implement this exactly.

It would be good to have some automated tests for such complicated cases because there is always a chance that by fixing one thing you break another.